chia/wallet/coin_selection.py

```diff
@@ -0,0 +1,53 @@
+import random
+from typing import Set, Optional, Tuple, List
+from chia.types.blockchain_format.coin import Coin
+from chia.util.ints import uint64
+from chia.wallet.wallet_coin_record import WalletCoinRecord
+
+
+# we use this to check if one of the coins exactly matches the target.
+def check_for_exact_match(coin_records: List[WalletCoinRecord], target: int) -> Optional[WalletCoinRecord]:
+    for coinrecord in coin_records:
+        if coinrecord.coin.amount == target:
+            return coinrecord
+    return None
+
+
+# we use this to find the individual coin that is closest to the target amount.
+def find_smallest_coin(coin_records: List[WalletCoinRecord], target: int) -> Optional[WalletCoinRecord]:
+    smallest_value = float("inf")  # smallest coins value
+    smallest_coin: Optional[WalletCoinRecord] = None
+    for coinrecord in coin_records:
+        if target < coinrecord.coin.amount < smallest_value:
+            # try to find a coin that is as close as possible to the amount.
+            smallest_value = coinrecord.coin.amount
+            smallest_coin = coinrecord
+    return smallest_coin
+
+
+# we use this to find the smallest set of coins.
+def knapsack_coin_algorithm(smaller_coins: Set[Coin], target: int) -> Tuple[Optional[Set[Coin]], uint64]:
+    best_set_sum = float("inf")
+    best_set_of_coins: Optional[Set[Coin]] = None
+    for i in range(1000):
+        # reset these variables every loop.
+        selected_coins: Set = set()
+        selected_coins_sum = 0
+        n_pass = 0
+        target_reached = False
+        while n_pass < 2 and not target_reached:
+            for coin in smaller_coins:

```

Comment:
You cannot iterate through a set, you will have to do `.items()` I think

Reply:
You can if i am not mistaken. https://www.geeksforgeeks.org/iterate-over-a-set-in-python/